### PR TITLE
Move away from Chan to STM

### PR DESCRIPTION
**Pull request by bgamari**
The isEmptyChan broken has been long known to be broken. Moreover, the atomicity guarantees provided by STM simplify things a bit.

### CODE REVIEW

**Comment by bgamari** [comment]
Note that this still needs to be testing on Windows.

**Comment by hvr** [comment]
@bgamari btw, a side-effect of this is that GHC would start including `stm` in its global package db!

**Comment by bgamari** [comment]
Oh no! Yes, I suppose you are right. That is quite unfortunate but I'm not sure I see an alternative. I suppose we *could* implement a locked queue with the semantics, but this doesn't seem like that much better of an alternative.

**Comment by hvr** [comment]
@bgamari as I said already on IRC, I think we should invest some time during the GHC 8.3.x dev cycle to figure out a way to decouple haskeline/terminfo from GHC's global pkg-db again so that haskeline can pick up new libraries (within reason) more easily :-)

**Comment by judah** [comment]
Thanks for the patch!

@hvr, I agree that decoupling from the global package DB would be a big help.  As another example, it currently limits which version of Haskeline can be included in a given Stackage LTS release.

Since it won't be necessary until ghc-8.4, and is a non-trivial change, I'll hold off on this PR until after we've cut the Haskeline release corresponding to ghc-8.2.1.

**Comment by bgamari** [comment]
Perhaps we could merge this now?

**Comment by judah** [comment]
@bgamari, LGTM except it looks like the PR is failing on Travis CI, I think due to missing a dependency on `stm` in `haskeline.cabal`.  Can you add that and try again?

**Comment by bgamari** [comment]
I think this should be ready now.

File: haskeline.cabal [hunk 1]
```diff
@@ -43,10 +43,10 @@ Library
     -- We require ghc>=7.4.1 (base>=4.5) to use the base library encodings, even
     -- though it was implemented in earlier releases, due to GHC bug #5436 which
     -- wasn't fixed until 7.4.1
-    Build-depends: base >=4.5 && < 4.11, containers>=0.4 && < 0.6,
+    Build-depends: base >=4.5 && < 4.12, containers>=0.4 && < 0.6,
                    directory>=1.1 && < 1.4, bytestring>=0.9 && < 0.11,
                    filepath >= 1.2 && < 1.5, transformers >= 0.2 && < 0.6,
-                   process >= 1.0 && < 1.7
+                   process >= 1.0 && < 1.7, stm >= 2.4 && < 2.6
```
Comment (hvr) [inline haskeline.cabal]: @bgamari did you intend to state  `< 2.5` here?

**Pull request merged**